pkg/signup/service/signup_service.go

```diff
@@ -333,30 +333,83 @@ func (s *ServiceImpl) reactivateUserSignup(ctx *gin.Context, existing *toolchain
 }
 
-func (s *ServiceImpl) DoGetSignup(provider ResourceProvider, userID, username string) (*signup.Signup, error) {
-	// Retrieve UserSignup resource from the host cluster, using the specified UserID and username
-	userSignup, err := s.DoGetUserSignupFromIdentifier(provider, userID, username)
-	// If an error was returned, then return here
-	if err != nil {
-		if apierrors.IsNotFound(err) {
+func (s *ServiceImpl) DoGetSignup(ctx *gin.Context, provider ResourceProvider, userID, username string) (*signup.Signup, error) {
+	var userSignup *toolchainv1alpha1.UserSignup
+	var err error
+
+	attempts := 0
+
+	for {
+		// If the loop exceeds a threshold in retry counts, then just return an error here
+		if attempts > 5 {
+			return nil, errs.New(fmt.Sprintf("failed to update UserSignup [%s] with updated userID or accountID: %s",
+				userSignup.Name,
+				userSignup.Status))
+		}
+
+		// Retrieve UserSignup resource from the host cluster, using the specified UserID and username
+		userSignup, err = s.DoGetUserSignupFromIdentifier(provider, userID, username)
+		// If an error was returned, then return here
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, nil
+			}
+			return nil, err
+		}
+
+		// Otherwise if the returned userSignup is nil, return here also
+		if userSignup == nil {
 			return nil, nil
 		}
-		return nil, err
-	}
 
-	// Otherwise if the returned userSignup is nil, return here also
-	if userSignup == nil {
-		return nil, nil
+		// Check the user_id and account_id annotations in the retrieved UserSignup.  If either of them are empty, but the
+		// values exist within the claims of the current user's Access Token then set the values in the UserSignup and update
+		// the resource.
+		userIDValue, userIDFound := userSignup.Annotations[toolchainv1alpha1.SSOUserIDAnnotationKey]
+		accountIDValue, accountIDFound := userSignup.Annotations[toolchainv1alpha1.SSOAccountIDAnnotationKey]
+
+		updateUserSignup := false
+
+		if !userIDFound || userIDValue == "" {
+			userID := ctx.GetString(context.UserIDKey)
+			if userID != "" {
+				userSignup.Annotations[toolchainv1alpha1.SSOUserIDAnnotationKey] = userID
+				updateUserSignup = true
+			}
+		}
+
+		if !accountIDFound || accountIDValue == "" {
+			accountID := ctx.GetString(context.AccountIDKey)
+			if accountID != "" {
+				userSignup.Annotations[toolchainv1alpha1.SSOAccountIDAnnotationKey] = accountID
+				updateUserSignup = true
+			}
+		}
+
+		// If there is no need to update the UserSignup then break out of the loop here
+		if !updateUserSignup {
+			break
+		}
+
+		userSignup, err = s.UpdateUserSignup(userSignup)
+		if err == nil {
+			// If there was no error during update, then break out of the loop here
+			break
+		}
+
+		// If we've reached this point, it means there was an error while attempting to update the UserSignup.  The
+		// attempts counter will be incremented and the for-loop will repeat
+		attempts++
 	}
```

Comment:
could you please reuse the already existing logic from:
https://github.com/codeready-toolchain/registration-service/blob/1c1f3bc60765ae8cf377ad9fdea1becd76740492/pkg/verification/service/verification_service.go#L426-L453

an example of how it is already used: https://github.com/codeready-toolchain/registration-service/blob/1c1f3bc60765ae8cf377ad9fdea1becd76740492/pkg/verification/service/verification_service.go#L163-L187

Reply:
I've copied the logic from the verification service and am now updating in a similar way.